File: crazyflie_lab/crazyflie_amr_py/src/pid_tuner.py
#!/usr/bin/env python

import numpy as np

import rclpy
from rclpy.node import Node
from rclpy.task import Future
from rclpy.qos import QoSProfile, QoSReliabilityPolicy
from geometry_msgs.msg import Twist

from crazyflie_interfaces.srv import *
from crazyflie_interfaces.msg import LogDataGeneric

from rcl_interfaces.srv import *
from rcl_interfaces.msg import *


START_Kp = 10.0
START_Kd = 0.0

class PIDTunerNode(Node):
    def __init__(self, cf_name = 'cf80'):
        super().__init__('pid_tuner')
        self.cf = cf_name

        qos_profile    = QoSProfile(depth=10, reliability=QoSReliabilityPolicy.BEST_EFFORT)
        self.gyro_sub  = self.create_subscription(LogDataGeneric, f"/{self.cf}/gyro_data", self.gyro_callback, qos_profile)
        self.accel_sub = self.create_subscription(LogDataGeneric, f"/{self.cf}/accel_data", self.accel_callback, qos_profile)
        self.vel_sub   = self.create_subscription(Twist, f"/{self.cf}/cmd_vel_legacy", self.vel_callback, qos_profile)
        
        self.timer_period_s = 1.0
        self.setup_timer    = self.create_timer(self.timer_period_s, self.setup)
        self.cmd_timer      = None

        # Variables for Node
        self.pitch_kp = START_Kp
        self.pitch_kd = START_Kd

        self.vel_active = False

        # Clients to get/set parameters
        param_names = [
            f"{self.cf}.params.pid_rate.pitch_kp", 
            f"{self.cf}.params.pid_rate.pitch_kd",
            f"{self.cf}.params.pid_rate.pitch_ki"]

        self.kp_idx = 0
        self.kd_idx = 1
        self.ki_idx = 2

        self.set_client       = self.create_client(SetParameters, '/crazyflie_server/set_parameters')
        self.set_msg          = SetParameters.Request()
        self.param_set_params = [Parameter(name=nm, value=ParameterValue(type=ParameterType.PARAMETER_DOUBLE)) for nm in param_names]
        self.set_future       = None

        # ------------------------------------- #
        # Optional: Add variables you want here #
                
        # ------------------------------------- #
        

    def setup(self):
        if self.set_client.service_is_ready():
            self.get_logger().warn("Parameter service ready!")

            self.param_set_params[self.kp_idx].value.double_value = self.pitch_kp
            self.param_set_params[self.kd_idx].value.double_value = self.pitch_kd
            self.param_set_params[self.ki_idx].value.double_value = 0.0

            self.set_msg.parameters = self.param_set_params
            self.set_future         = self.set_client.call_async(self.set_msg)

            self.setup_timer.cancel()
            self.cmd_timer = self.create_timer(self.timer_period_s, self.control_func)

            # --------------------------------------- #
            # Optional: Additional Setup actions here #

            # --------------------------------------- #
        else:
            self.get_logger().info("Parameter service NOT ready")

    def control_func(self):
        if self.ang_vel_received and self.accel_received and self.vel_active:
            self.ang_vel_received = False
            self.accel_received   = False

            # First, checking if call to set parameter has finished
            if self.set_future:
                if self.set_future.done():
                    response = self.set_future.result()
                    if response is not None:
                        for rst in response.results:
                            if not rst.successful:
                                self.get_logger().warn(f"Failed to set parameter. Reason: {rst.reason}")
                    else:
                        self.get_logger().warn("Call to set parameter service failed")
                    
                    self.set_future = None
                    
            # If call has finished, future will be set to "None"
            # Update gains and set parameters again
            if not self.set_future:
                # ---------------------------- #
                #    Your logic goes here      #

                # Example: Setting proportional and derivative gains
                self.pitch_kp = 0.0
                self.pitch_kd = 0.0
    
                # ---------------------------  #

                # Call to service to set parameters
                self.param_set_params[self.kp_idx].value.double_value = self.pitch_kp
                self.param_set_params[self.kd_idx].value.double_value = self.pitch_kd

                self.set_msg.parameters = self.param_set_params
                self.set_future         = self.set_client.call_async(self.set_msg)
 

    def gyro_callback(self, msg):
        self.get_logger().info(f"Ang. Velocity X: {msg.values[0]}  deg/s")
        self.get_logger().info(f"Ang. Velocity Y: {msg.values[1]}  deg/s")
        self.get_logger().info(f"Ang. Velocity Z: {msg.values[2]}  deg/s")

        # ------------------------------------------ #
        #    Store angular velocity data [deg/s]     #

        # ------------------------------------------ #        

    def accel_callback(self, msg):
        self.get_logger().info(f"Acceleration X: {msg.values[0]}  deg/s")
        self.get_logger().info(f"Acceleration Y: {msg.values[1]}  deg/s")
        self.get_logger().info(f"Acceleration Z: {msg.values[2]}  deg/s")

        # ------------------------------------------ #
        #       Store accleration data [g]           #

        # ------------------------------------------ #    

    def vel_callback(self, msg):
        self.vel_active = msg.linear.z > 0.0
        
def main(args=None):
    rclpy.init(args=args)
    
    pid_tune_node = PIDTunerNode()
    rclpy.spin(pid_tune_node)

    pid_tune_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: crazyflie_lab/crazyflie_amr_cpp/src/pid_tuner.cpp
#include <rclcpp/rclcpp.hpp>
#include <rmw/qos_profiles.h>

#include <rcl_interfaces/srv/set_parameters.hpp>
#include <rcl_interfaces/msg/parameter.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <crazyflie_interfaces/msg/log_data_generic.hpp>

#include <chrono>
#include <algorithm>

using namespace std::chrono_literals;

class PidTunerNode : public rclcpp::Node
{
public:
  ~PidTunerNode(void) = default;

  PidTunerNode(const std::string &cf_name="cf80") 
    : Node("pid_tuner"),
    _cf_name(cf_name),
    _vel_cmd_active(false),
    _ang_vel_received(false)
  {
    auto qos_profile = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_sensor_data), rmw_qos_profile_sensor_data);

    _vel_sub = this->create_subscription<geometry_msgs::msg::Twist>(
      "/" + cf_name + "/cmd_vel_legacy", qos_profile,
      [this](const geometry_msgs::msg::Twist &msg) { this->vel_callback(msg); });

    _gyro_sub = this->create_subscription<crazyflie_interfaces::msg::LogDataGeneric>(
      "/" + cf_name + "/gyro_data", qos_profile,
      [this](const crazyflie_interfaces::msg::LogDataGeneric &msg) { this->gyro_callback(msg); });

    _accel_sub = this->create_subscription<crazyflie_interfaces::msg::LogDataGeneric>(
      "/" + cf_name + "/accel_data", qos_profile,
      [this](const crazyflie_interfaces::msg::LogDataGeneric &msg) { this->accel_callback(msg); });

    _setup_timer = this->create_wall_timer(500ms, [this](void) { this->setup(); });

    _set_param_client = this->create_client<rcl_interfaces::srv::SetParameters>("/crazyflie_server/set_parameters");
  }

private:

  void setup(void) 
  {
    if(_set_param_client->service_is_ready()) {
      RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Set Parameter Service is ready.");

      _pitch_Kp = kStartKp;
      _pitch_Kd = kStartKd;

      rcl_interfaces::msg::Parameter kp;
      kp.name               = _cf_name + ".params.pid_rate.pitch_kp";
      kp.value.type         = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
      kp.value.double_value = _pitch_Kp;

      rcl_interfaces::msg::Parameter kd;
      kd.name               = _cf_name + ".params.pid_rate.pitch_kd";
      kd.value.type         = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
      kd.value.double_value = _pitch_Kd;

      rcl_interfaces::msg::Parameter ki;
      ki.name               = _cf_name + ".params.pid_rate.pitch_ki";
      ki.value.type         = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
      ki.value.double_value = 0.0;

      _set_param_req = std::make_shared<rcl_interfaces::srv::SetParameters::Request>();
      _set_param_req->parameters = std::vector<rcl_interfaces::msg::Parameter>({kp, kd, ki});
      _set_param_result = _set_param_client->async_send_request(_set_param_req).share();

      _cmd_timer = this->create_wall_timer(1000ms, [this](void) { this->control_func(); });
      _setup_timer->cancel();

      // --------------------------- //
      // Optional setup actions here //

      // --------------------------- //
    } else {
      RCLCPP_WARN(rclcpp::get_logger("pid_tuner"), "Set Parameter Service is NOT ready.");
    }
  }

  void control_func(void)
  {
    if(_set_param_result.valid()) {
      if(_set_param_result.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
        _set_param_resp = _set_param_result.get();
        std::vector<rcl_interfaces::msg::SetParametersResult> results =  _set_param_resp->results;

        // Reset the future so that valid will return false
        _set_param_result = rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedFuture();
      }
    }

    // If call has finished, update gains and set parameters again
    if(!_set_param_result.valid() && _vel_cmd_active) {
      // --------------------------- //
      //    Your logic goes here     //

    
      // Example: Setting proportional and derivative gains
      _pitch_Kp = 0.0;
      _pitch_Kd = 0.0;

      // ---------------------------  //

      // Call to set parameters service
      _set_param_req->parameters[0].value.double_value = _pitch_Kp;
      _set_param_req->parameters[1].value.double_value = _pitch_Kd;

      _set_param_result = _set_param_client->async_send_request(_set_param_req).share();
    }
  }

  void gyro_callback(const crazyflie_interfaces::msg::LogDataGeneric &msg)
  {
    RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Ang Velocity X: %.2f deg/s", msg.values[0]);
    RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Ang Velocity Y: %.2f deg/s", msg.values[1]);
    RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Ang Velocity Z: %.2f deg/s", msg.values[2]);


    // ------------------------------------------ //
    //    Store angular velocity data [deg/s]     //

    // ------------------------------------------ //
  }

  void accel_callback(const crazyflie_interfaces::msg::LogDataGeneric &msg)
  {
    RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Accel X: %.2f g", msg.values[0]);
    RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Accel Y: %.2f g", msg.values[1]);
    RCLCPP_INFO(rclcpp::get_logger("pid_tuner"), "Accel Z: %.2f g", msg.values[2]);

    // ------------------------------------------ //
    //          Store accleration data [g]        //
    

    // ------------------------------------------ //
  }

  void vel_callback(const geometry_msgs::msg::Twist &msg) 
  {
    _vel_cmd_active = msg.linear.z > 0.0;
  }


  // Variables used by the Node Object //

  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr _vel_sub;
  rclcpp::Subscription<crazyflie_interfaces::msg::LogDataGeneric>::SharedPtr _gyro_sub;
  rclcpp::Subscription<crazyflie_interfaces::msg::LogDataGeneric>::SharedPtr _accel_sub;
  rclcpp::TimerBase::SharedPtr _cmd_timer;
  rclcpp::TimerBase::SharedPtr _setup_timer;

  rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedPtr    _set_param_client;
  rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedFuture _set_param_result;
  std::shared_ptr<rcl_interfaces::srv::SetParameters::Request>     _set_param_req;
  std::shared_ptr<rcl_interfaces::srv::SetParameters::Response>    _set_param_resp;

  std::string _cf_name;

  bool _vel_cmd_active;

  double _pitch_Kp;
  double _pitch_Kd;

  static constexpr double kStartKp = 0.0;
  static constexpr double kStartKd = 0.0;

  // ------------------------------------- //
  // Optional: Add variables you want here //

  // ------------------------------------- //
};


int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PidTunerNode>("cf80"));
  rclcpp::shutdown();

  return 0;
}